relayer/chains/substrate/substrate_chain_processor.go

```diff
@@ -0,0 +1,379 @@
+			scp.log.Error("Found channel using multiple connection hops. Not currently supported, ignoring.",
+				zap.String("channel_id", ch.ChannelId),
+				zap.String("port_id", ch.PortId),
+				zap.Strings("connection_hops", ch.ConnectionHops),
+			)
+			continue
+		}
+		scp.channelConnections[ch.ChannelId] = ch.ConnectionHops[0]
+		scp.channelStateCache[processor.ChannelKey{
+			ChannelID:             ch.ChannelId,
+			PortID:                ch.PortId,
+			CounterpartyChannelID: ch.Counterparty.ChannelId,
+			CounterpartyPortID:    ch.Counterparty.PortId,
+		}] = ch.State == chantypes.OPEN
+	}
+	return nil
+}
+
+func (scp *SubstrateChainProcessor) queryCycle(ctx context.Context, persistence *queryCyclePersistence) error {
+	var err error
+	persistence.latestHeight, err = scp.latestHeightWithRetry(ctx)
+
+	// don't want to cause CosmosChainProcessor to quit here, can retry again next cycle.
+	if err != nil {
+		scp.log.Error(
+			"Failed to query latest height after max attempts",
+			zap.Uint("attempts", latestHeightQueryRetries),
+			zap.Error(err),
+		)
+		return nil
+	}
+
+	scp.log.Debug("Queried latest height",
+		zap.Int64("latest_height", persistence.latestHeight),
+	)
+
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
+
+	if !scp.inSync {
+		if (persistence.latestHeight - persistence.latestQueriedBlock) < inSyncNumBlocksThreshold {
+			scp.inSync = true
+			firstTimeInSync = true
+			scp.log.Info("Chain is in sync")
+		} else {
+			scp.log.Info("Chain is not yet in sync",
+				zap.Int64("latest_queried_block", persistence.latestQueriedBlock),
+				zap.Int64("latest_height", persistence.latestHeight),
+			)
+		}
+	}
+
+	ibcMessagesCache := processor.NewIBCMessagesCache()
+
+	ibcHeaderCache := make(processor.IBCHeaderCache)
+
+	ppChanged := false
+
+	var latestHeader SubstrateIBCHeader
+
+	newLatestQueriedBlock := persistence.latestQueriedBlock
+
+	for i := persistence.latestQueriedBlock + 1; i <= persistence.latestHeight; i++ {
+		var eg errgroup.Group
+		var ibcEvents rpcclienttypes.IBCEventsQueryResult
+		var ibcHeader provider.IBCHeader
+		i := i
+		heightUint64 := uint64(i)
+
+		eg.Go(func() (err error) {
+			queryCtx, cancelQueryCtx := context.WithTimeout(ctx, queryTimeout)
+			defer cancelQueryCtx()
+			ibcEvents, err = scp.chainProvider.RPCClient.RPC.IBC.QueryIbcEvents(queryCtx, []rpcclienttypes.BlockNumberOrHash{{Number: uint32(heightUint64)}})
+			return err
+		})
+		eg.Go(func() (err error) {
+			queryCtx, cancelQueryCtx := context.WithTimeout(ctx, queryTimeout)
+			defer cancelQueryCtx()
+			ibcHeader, err = scp.chainProvider.QueryIBCHeader(queryCtx, i)
+			return err
+		})
+
+		if err := eg.Wait(); err != nil {
+			scp.log.Warn("Error querying block data", zap.Error(err))
+			break
+		}
+
+		latestHeader = ibcHeader.(SubstrateIBCHeader)
+
+		// TODO: update latest bolock
```

Comment:
Should this still be commented out? After comparing the Substrate processor with the Cosmos processor, it looks like this is the only piece missing.

Reply:
Done in the [PR#986](https://github.com/cosmos/relayer/pull/986)